Exercise: calculator

listochkin opened the pull request:
Port of an older exercise with reworked structure that addresses issues I've been running into with groups in past two years.

The help text and a step by step solution promote the use of `?` as a default error handling option.

A step by step solution also mentions Rust Analyzer commands that people new to Rust would greatly benefit from.

listochkin committed:
Exercise: calculator

Port of an older exercise with reworked structure that addresses issues I've been running into with groups in past two years.

The help text and a step by step solution promote the use of `?` as a default error handling option.

A step by step solution also mentions Rust Analyzer commands that people new to Rust would greatly benefit from.

jonathanpallant commented on exercise-book/src/calculator.md:
```diff
@@ -0,0 +1,700 @@
+Write a library that parses and evaluates math expressions that use [Reverse-Polish (postfix) notation](https://en.wikipedia.org/wiki/Reverse_Polish_notation).
+For example, a string `"3 1 + 2 /"` in this notation is equivalent to `(3 + 1) / 2` but unlike the later it does not force us to handle operator precedence using parenthesis.
+
+We will support the 4 basic math operations (`+`, `-`, `*`, `/`) that all will expect two operants, and a `sqr` operation to square a single number.
```
```suggestion
We will support the 4 basic math operations (`+`, `-`, `*`, `/`) that all will expect two operands, and a `sqr` operation to square a single number.
```

jonathanpallant commented on exercise-book/src/calculator.md:
```diff
@@ -0,0 +1,700 @@
+    for word in input.split_ascii_whitespace() {
+
+    };
+    assert_eq!(stack.len(), 1);
```
I'm not sure that suggesting `unwrap()` by default is the best solution here, but I'll go with it if you're sure it's best as I haven't done this exercise.

listochkin replied:
This actually is a leftover from old times. I've put the starter code as is, and the exercise Help section talks about this `assert_eq!` and the `unwrap` are problematic.

The participants run into it often, and some get confused how their tests are visibly failing, but the error message is not what they have in the test body and instead is something completely unrelated. I turn it into a teachable moment instead, and with this exercise rewrite one of the steps in the solution rewrite this portion of code, too.

jonathanpallant commented on exercise-book/src/calculator.md:
```diff
@@ -0,0 +1,700 @@
+
+Some observations:
+
+* Every branch of our big `match` statement eds up producing an expression.
```
```suggestion
* Every branch of our big `match` statement ends up producing an expression.
```

jonathanpallant commented on exercise-book/src/calculator.md:
```diff
@@ -0,0 +1,700 @@
+match expr {
+    // ...
+    Expr::Add(a, b) => eval(a)? + eval(b)?,
+    Expr::Sub(a, b) => eval(a)? - eval(b)?,
```
Should we talk about `checked_add` and `checked_sub` and how arithmetic can panic and so shouldn't be performed on untrusted inputs?

listochkin replied:
Probably? When I'm running this exercise with groups of different backgrounds everyone remembers the division by zero, but only people with some C or embedded background mention overflows (in three years it only happened once!). So, I keep it light and simple with operators. 

Handling overflows can be a stretch goal, and I can mention it in the exercise. Good call!

jonathanpallant reviewed (commented):
Couple of spelling nits, and a few questions, but otherwise good to go. Very happy to see this arrive!

listochkin committed:
Calculator exercise: add stretch goals

I decided not to change the solution.
The stretch goals would probably be done as homework.
